libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -77,32 +77,23 @@ init_pltgot(Obj_Entry *obj)
  * 8-bit permission field.
  */
 static void
-init_cap_from_fragment(caddr_t relocbase, const void *text_rodata_cap,
-    Elf_Addr fragment_offset, Elf_Size addend)
+init_cap_from_fragment(void *where, void * __capability data_cap,
+    const void * __capability text_rodata_cap, Elf_Addr base_addr,
+    Elf_Size addend)
 {
-	void *fragment;
+	Elf_Addr *fragment;
 	uintcap_t cap;
-	uint64_t address, len;
+	Elf_Addr address, len;
 	uint8_t perms;
 
-	fragment = (void *)(relocbase + fragment_offset);
-	address = *((uint64_t *)fragment);
-	len = *((uint64_t *)fragment + 1) & 0xffffffffffffffULL;
-	perms = *((uint64_t *)fragment + 1) >> 56;
+	fragment = (Elf_Addr *)where;
```

Comment:
I feel like this would be a bit easier to read if we define a struct for the fragment?

Reply:
Yeah I was thinking about whether it's worth it or not. I guess it helps to avoid the duplication of the size calculation for the length and perms fields.

Reply:
I guess if we use bitfields we depend on compiler-specific behaviour, and a struct of two uint64_t doesn't help much.

Reply:
Yes, although the fragment format is already tied to the ABI so we can rely on other parts of the ABI too. Plus we're never going to use that format for our CHERI implementations, we'll just splat the raw capability into memory at link time, do the CRelocate + CBuildCap (presumably with two different relocations to determine which root to use, unless we do a CCBuildCap chain with multiple roots) and write it back.

Reply:
Actually I think I'd rather not; whilst it works for arm64, you can't have a definition that would allow cross tools to use it due to endianness issues with bitfields unless you `#ifdef` and I don't want to have to remember exactly how bitfields work with endianness. We can always add it later if we decide we really want it (e.g. if we end up needing the struct outside of rtld).